metaserver/mongoDB.go

```diff
@@ -0,0 +1,471 @@
+	c, session, err := getMongoCollection("providers")
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	selector := struct{ ID string }{ID: providerID}
+	var result core.ProviderInfo
+	err = c.Find(selector).One(&result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// Insert the given provider into the database.
+func (db *mongoDB) InsertProvider(provider core.ProviderInfo) error {
+	c, session, err := getMongoCollection("providers")
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	err = c.Insert(provider)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Update the given provider in the databse.
+func (db *mongoDB) UpdateProvider(provider core.ProviderInfo) error {
+	c, session, err := getMongoCollection("providers")
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	selector := struct{ ID string }{ID: provider.ID}
+	err = c.Update(selector, provider)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete the specified provider from the dtabase.
+func (db *mongoDB) DeleteProvider(providerID string) error {
+	c, session, err := getMongoCollection("providers")
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	selector := struct{ ID string }{ID: providerID}
+	err = c.Remove(selector)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// File operations
+//====================
+
+// Return a list of all files in the database.
+func (db *mongoDB) FindAllFiles() ([]core.File, error) {
+	c, session, err := getMongoCollection("files")
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	var result []core.File
+	err = c.Find(nil).All(&result)
+
+	return result, nil
+}
+
+// Return the latest version of the file with the specified ID.
+func (db *mongoDB) FindFileByID(fileID string) (*core.File, error) {
+	c, session, err := getMongoCollection("files")
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	selector := struct{ ID string }{ID: fileID}
+	var result core.File
+	err = c.Find(selector).One(&result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// Return a map of paths to files present in the renter's directory.
+func (db *mongoDB) FindFilesInRenterDirectory(renterID string) ([]core.File, error) {
+	session, err := mgo.Dial(dbAddress)
+	if err != nil {
+		return nil, err
+	}
+	defer session.Close()
+
+	// Get file IDs in renter directory.
+	renters := session.DB(dbName).C("renters")
+
+	selector := struct{ ID string }{ID: renterID}
+	var renter core.RenterInfo
+	err = renters.Find(selector).One(&renter)
+	if err != nil {
+		return nil, err
+	}
+	filesToFind := renter.Files
+
+	// Retrieve files from collection.
+	files := session.DB(dbName).C("files")
+
+	var foundFiles []core.File
+	for _, item := range filesToFind {
+		selector = struct{ ID string }{ID: item}
+		var result core.File
+		err = files.Find(selector).One(&result)
```

Comment:
Does Mongo support joins?

Reply:
Looks like it: https://docs.mongodb.com/master/reference/operator/aggregation/lookup/#pipe._S_lookup

I totally missed it my first time glancing through the docs. I'll switch it to use that.

Reply:
I'm going to leave this one for later, since it is a lot more complicated than what we're doing now. I'll revisit it after the alpha demo.